infrastructure/async/src/main/java/tech/pegasys/teku/infrastructure/async/SafeFuture.java

```diff
@@ -235,6 +235,41 @@ public static <U> SafeFuture<Void> allOfFailFast(final SafeFuture<?>... futures)
     return complete;
   }
 
+  /**
+   * Returns a new {@link SafeFuture} with the result of the first successful future from the given
+   * futures. If all futures complete exceptionally, the returned {@link SafeFuture} will complete
+   * exceptionally with the exception from the first future. If the provided {@link List} of futures
+   * is an empty collection, the returned {@link SafeFuture} will complete exceptionally with an
+   * {@link IllegalArgumentException} exception.
+   *
+   * @param futures the futures which will need to be completed
+   * @return a new {@link SafeFuture} that is completed with the first successful result from the
+   *     given futures or completed exceptionally if all futures complete exceptionally
+   */
+  public static <T> SafeFuture<T> firstSuccess(final List<SafeFuture<T>> futures) {
+    if (futures.isEmpty()) {
+      return SafeFuture.failedFuture(
+          new IllegalArgumentException("The provided list of SafeFuture<T> should not be empty"));
+    }
+    return asyncDoWhile(
+            () ->
+                SafeFuture.of(
+                    () -> {
+                      final boolean canProceed =
+                          futures.stream().anyMatch(SafeFuture::isCompletedNormally)
+                              || futures.stream().allMatch(SafeFuture::isCompletedExceptionally);
+                      return !canProceed;
+                    }))
+        .thenCompose(
+            __ ->
+                futures.stream()
+                    .filter(SafeFuture::isCompletedNormally)
+                    .findFirst()
+                    // at this point, all futures have completed exceptionally, so just
+                    // return the first error future
+                    .orElse(futures.get(0)));
```

Comment:
We need to add tests for this method. :)

I don't think `asyncDoWhile` is going to do what you expect here - I think this will wind up in a tight loop continuously executing while the futures execute instead of actually waiting for the next future to complete before looping again because the `SafeFuture.of` here is immediately returning. 

I think you can simplify it to just:
```java
    final SafeFuture<T> result = new SafeFuture<>();
    allOf(
            futures.stream()
                .map(future -> future.thenAccept(result::complete))
                .toArray(SafeFuture[]::new))
        .propagateExceptionTo(result);
    return result;
```

which takes advantage of the fact that you can only complete a future once so for each future we tell it to pass it's successful result to the returned future as soon as it finishes, then once all the futures are complete, propagate any exception. Whatever completes the `result` future first is the result that will actually be used and everything after that is just ignored.

There will likely be a warning about unchecked types to suppress because the `SafeFuture[]::new` but that's fine.